src/js/plugins/plugin.search.js

```diff
@@ -0,0 +1,575 @@
+          <div class="loader tc mt20"></div>`
+      );
+      return false;
+    };
+
+    // Bind search forms
+    this.$('.BRbooksearch.desktop').submit(onSubmitDesktop);
+    this.$('.BRbooksearch.mobile').submit(onSubmitMobile);
+
+    // Handle clearing the search results
+    this.$(".BRsearchInput").bind('input propertychange', () => {
+      if (this.value === "") this.removeSearchResults();
+    });
+  };
+})(BookReader.prototype.initToolbar);
+
+
+/**
+ * Submits search request
+ *
+ * @param { string } term
+ * @param { object } options
+ * @param { boolean } options.disablePopup
+ */
+BookReader.prototype.search = function(term, options) {
+  options = options !== undefined ? options : {};
+  const defaultOptions = {
+    /**
+     * @type { bool } - (default=false) goToFirstResult - jump to the first result
+     */
+    goToFirstResult: false,
+    /**
+     * @type { bool } - (default=false) disablePopup - don't show the modal progress
+     */
+    disablePopup: false
+  };
+  options = jQuery.extend({}, defaultOptions, options);
+
+  this.$('.BRsearchInput').blur(); //cause mobile safari to hide the keyboard
+
+  this.removeSearchResults();
+
+  this.searchTerm = term;
+  // strip slashes, since this goes in the url
+  this.searchTerm = this.searchTerm.replace(/\//g, ' ');
+
+  this.trigger(BookReader.eventNames.fragmentChange);
+
+  // Add quotes to the term. This is to compenstate for the backends default OR query
+  // term = term.replace(/['"]+/g, '');
+  // term = '"' + term + '"';
+
+  // Remove the port and userdir
+  const serverPath = this.server.replace(/:.+/, '');
+  const baseUrl = `https://${serverPath}${this.searchInsideUrl}?`;
+
+  // Remove subPrefix from end of path
+  let path = this.bookPath;
+  const subPrefixWithSlash = `/${this.subPrefix}`;
+  if (this.bookPath.length - this.bookPath.lastIndexOf(subPrefixWithSlash) == subPrefixWithSlash.length) {
+    path = this.bookPath.substr(0, this.bookPath.length - subPrefixWithSlash.length);
+  }
+
+  const urlParams = {
+    'item_id': this.bookId,
+    'doc': this.subPrefix,
+    'path': path,
+    'q': term,
+  };
+
+  let paramStr = $.param(urlParams);
+
+  // NOTE that the API does not expect / (slashes) to be encoded. (%2F) won't work
+  paramStr = paramStr.replace(/%2F/g, '/');
+
+  const url = `${baseUrl}${paramStr}`;
+
+  if (!options.disablePopup) {
+    this.showProgressPopup(
+      `<img class="searchmarker" src="${this.imagesBaseURL}marker_srch-on.svg" />
+      Search results will appear below...`);
+  }
+
+  const processSearchResults = (data) => {
+    if (data.error || 0 == data.matches.length) {
+      this.BRSearchCallbackError(data, options);
+    } else {
+      this.BRSearchCallback(data, options);
+    }
+  };
+
+  $.ajax({
+    url: url,
+    dataType: 'jsonp'
+  }).then(processSearchResults);
+};
+
+/**
+ * Search Results return handler
+ * @param { object } results
+ * @param { array } results.matches
```

Comment:
Oh, last thing; let's give these some `typedef`s so that we can reference their types, and add info about what fields are present. I think that would help make things much clearer!

Reply:
yes I like. :excellent: